File: 910.最小差值-ii.py
#
# @lc app=leetcode.cn id=910 lang=python3
#
# [910] 最小差值 II
#

# @lc code=start
class Solution:
    def smallestRangeII(self, nums: List[int], k: int) -> int:
        nums.sort()
        length = len(nums)
        res = nums[length - 1] - nums[0]
        # 最差的情况==最小的加上k和最大的减去k
        p1 = nums[0] + k
        p4 = nums[length-1] - k
        # 有序，所以较小的加k 较大的-k 比较
        for i in range(length-1):
            p2 = nums[i] + k
            p3 = nums[i+1] - k
            h = max(p2, p4)
            l = min(p1, p3)
            res = min(res, h-l)
        return res
# @lc code=end

File: 910.最小差值-ii.go
/*
 * @lc app=leetcode.cn id=910 lang=golang
 *
 * [910] 最小差值 II
 */

// @lc code=start
func smallestRangeII(nums []int, k int) int {
	sort.Ints(nums)
	length := len(nums)
	res := nums[length-1] - nums[0]
	p1 := nums[0] + k
	p4 := nums[length-1] - k
	for i := 0; i < length-1; i++ {
		p2 := nums[i] + k
		p3 := nums[i+1] - k
		h := max(p2, p4)
		l := min(p1, p3)
		if res > h-l {
			res = h - l
		}
	}
	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x > y {
		return y
	}
	return x
}

// @lc code=end

